unittests/integration/draw_tester.cpp

```diff
@@ -0,0 +1,48 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "CUDAQTestUtils.h"
+#include <cudaq/algorithms/draw.h>
+
+CUDAQ_TEST(DrawTester, checkEmpty) {
+
+  auto kernel = []() __qpu__ {};
+
+  std::string expected_str = "<empty trace>";
+  auto produced_str = cudaq::draw(kernel);
+  EXPECT_EQ(expected_str, produced_str);
+}
+
+CUDAQ_TEST(DrawTester, checkOps) {
```

Comment:
We need more testing than that. 
- We need tests to check the error message if the kernel contains measurements (at the end or in the middle)
- We need to test the output if the kernel calls another kernel 
- We need to cover all combinations of adjoint and controlled (does ctrl adj and adj ctrl plot properly?)
- We need to cover all operations in the default set documented [here](https://nvidia.github.io/cuda-quantum/latest/api/default_ops.html)

Reply:
> We need more testing than that.

These tests capture few edge cases that can happen in the drawer. I will add the tests you are requesting, but they won't be testing the drawer, they will be testing the execution manager capability of creating a correct trace. (It turns out it does get lost in some cases, but maybe those tests shouldn't here)

> * We need tests to check the error message if the kernel contains measurements (at the end or in the middle)

As stated in the PR description: This restrictions is not yet implemented because the drawer is based on the tracing capabilities of the execution manager, and the tracer still allows measurements.

A later PR needs implement this restriction but it will be a breaking change, as the current tracer functionality is public facing. (Although, I have doubts of if anyone is using it.)

> * We need to test the output if the kernel calls another kernel

Just keep in mind that the execution manager does not "see" calls to other kernels. All it sees are call to operations. Furthermore, the tracer doesn't even see calls to qubit allocation and deallocation, so it must infer the number of qubits from operations---this will make results a bit unexpected in the drawer, but is a fix that needs to be done in the execution manager.

> * We need to cover all combinations of adjoint and controlled (does ctrl adj and adj ctrl plot properly?)

The current tracer does not discriminate `adj`. It receives only an operation name, which is used as the label, and a list of qudits (the execution manager modifies the name in the case of `adj`). Also, because `cudaq::adj` and `cudaq::ctrl` are types, one cannot combine `op<ctrl, adj>(...)` (a pain point we should address in the future)

Reply:
1) Could you please create a GitHub issue to capture the remaining work, including any test cases that still need to be added (minimally the ones I listed above).
2) I am looking for functionality tests - hence I do not care which piece of code ultimately supports or is changed to support that; the tests should not need to know that or rely on that. In the end, we want a certain behavior for the end-usage, and that's what I would like to see some tests for. It's fine to split that up across multiple PRs, but please capture all work in the same GitHub issue so that it is easier to track.
3) Re not seeing other kernels: Yes, I'm aware and that's fine - we'll make this fancier with and when we revise this more broadly. However, I'd still like to have a test now, since we still want that to work - just broken down into individual gates (kernel is effectively "inlined").
4) Having Adj be reflected in the gate name sounds fine.

Reply:
The tests are already testing all gates. They also test calling other kernels and a combination of `cudaq::control`, `cudaq::adjoint` and `op<cudaq::adj>(...)`. 

I opened an issues regarding running the tracer in the presence of measurements, and the fact that it does not capture all quantum operations, e.g., qubit allocation.